extensions/csrc/cuda/funcs/op_functor.h

```diff
@@ -1,31 +1,91 @@
 #pragma once
 
 #include <cuda.h>
+#include <cuda_bf16.h>
 #include <cuda_fp16.h>
 #include <cuda_runtime.h>
 
 #include <functional>
 
+#include "../utils/micros.h"
+
 namespace colossalAI {
 namespace cuda {
 namespace funcs {
 
-enum class BinaryOpType { kAdd = 0, kMinus, kMul, KDiv, kMax, KMin };
+enum class BinaryOpType { kAdd = 0, kMinus, kMul, kDiv, kMax, kMin };
 
-template <typename T, BinaryOpType Op>
+// Note(LiuYang): This file provides base math operation for data type
+// include POD and cuda built-in type such as half and __nv_bfloat16
+template <typename LT, typename RT, typename RET, BinaryOpType Op>
 struct BinaryOpFunctor;
 
-template <typename T>
-struct BinaryOpFunctor<T, BinaryOpType::kAdd>
-    : public std::binary_function<T, T, T> {
-  __host__ __device__ T operator()(T lhs, T rhs) { return lhs + rhs; }
-};
-
-template <typename T>
-struct BinaryOpFunctor<T, BinaryOpType::kMax>
-    : public std::binary_function<T, T, T> {
-  __host__ __device__ T operator()(T lhs, T rhs) { return max(lhs, rhs); }
-};
+#define COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BINARY_OP_TYPE, STMT,    \
+                                               FUNCTION_MODIFIER, ARGS...) \
+  template <ARGS>                                                          \
+  struct BinaryOpFunctor<T, T, T, BINARY_OP_TYPE>                          \
+      : public std::binary_function<T, T, T> {                             \
+    FUNCTION_MODIFIER T operator()(T lhs, T rhs) { return STMT; }          \
+  };
+
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BinaryOpType::kAdd, lhs + rhs,
+                                       HOSTDEVICE, typename T)
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BinaryOpType::kMinus, lhs - rhs,
+                                       HOSTDEVICE, typename T)
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BinaryOpType::kMul, lhs* rhs,
+                                       HOSTDEVICE, typename T)
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BinaryOpType::kDiv, lhs / rhs,
+                                       HOSTDEVICE, typename T)
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BinaryOpType::kMax, max(lhs, rhs),
+                                       HOSTDEVICE, typename T)
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(T, BinaryOpType::kMin, min(lhs, rhs),
+                                       HOSTDEVICE, typename T)
+
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(half, BinaryOpType::kAdd,
+                                       __hadd(lhs, rhs), DEVICE)
+COLOSSAL_BINARY_FUNCTOR_SPECIALIZATION(half2, BinaryOpType::kAdd,
+                                       __hadd2(lhs, rhs), DEVICE)
+
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 800
```

Comment:
Might want to add notes for CUDA arch support for bfloat16 operations and future plan or TODO.